Optimizations

andrewmd5 opened the pull request:
This PR addresses a few memory leaks and aims to reduce runtime allocations. Here are some of the changes:
- Await individual task directly rather than allocating multiple task into a single group. 
- Forward exceptions to OnException directly to avoid delegate allocations. 
- Because we know the max size of a chunk, I've Introduced a reusable StringBuilder to the DatadogHttpClient and LogFormatter so batches don't reallocate. 
- Add a lightweight semaphore to ensure only one batch of events is in-flight at any given time. (needed to allow for reusable builders)
- Adds a flowing cancellation token that will ensure resources get released when the sink is disposed, or if a client breaks.  
- Adds a delay to the sink dispose method to ensure pending batches get written.  
- Constructs chunks used by the HttpClient inline.
- Other minor fixes. 

Tested and haven't seen any issues. We run our .NET code in environments with very little memory (1 GB) and these changes help ensure logging isn't creating memory pressure.

AndrewMD5 commented:
One note - it would be possible to entirely remove the round-trip JSON de(serialization) that occurs in the formatter by leveraging the LogEvent properties. For instance, calling `logEvent.AddPropertyIfAbsent(new LogEventProperty("service", new ScalarValue(_service)));` would set the service under the properties object, which gets serialized like this: 
```json
{
      "Timestamp":"2021-11-01T15:52:29.6217246-07:00",
      "level":"Error",
      "MessageTemplate":"Yo!",
      "message":"Yo!",
      "Exception":"System.Exception: test 1",
      "Properties":{
         "service":"gateway"
      },
      "host":"DESKTOP-I1EFOH0"
   }
```

But this would be a breaking change that would require the Datadog backend to be updated.

gh123man commented on src/Serilog.Sinks.Datadog.Logs/Sinks/Datadog/DatadogHttpClient.cs:
```diff
@@ -36,123 +41,139 @@ public class DatadogHttpClient : IDatadogClient
+        /// <param name="onException">The callback that is invoked when an exception occurs in this event but is handled gracefully.</param>
+        /// <remarks>
+        /// <para>This method uses a semaphore so that only one batch of events is ever on the wire at any given time.</para>
+        /// <para>This allows for a single allocated contiguous memory to be written to for serialization of events.</para>
```
Thanks for introducing this. 
Would it be possible to put this behind a config flag? In scenarios where apps are sending a very high volume of logs someone may prioritize concurrent throughput over memory usage. The round trip latency of posting a payload to the intake and waiting for it to complete can impose a cap on how fast logs can be sent.

AndrewMD5 replied:
Good catch. I can make that change.

AndrewMD5 commented:
I've put the resource recycling behind a flag on the configuration object. I've also gone ahead and removed the dependency on System.Text.Json and Newtonsoft.Json. I implemented a custom Serilog formatter which forks the built in JSON formatter and writes data in the format the original implementation was trying to achieve without the roundtrip serialization.

AndrewMD5 commented:
@gh123man I was also curious- once this is merged is it possible to get a new release on Nuget? Even if it’s pre-release that would work for us.

gh123man commented:
@AndrewMD5  Yep sure thing - ill publish a release once this goes in.

gh123man reviewed (dismissed):
This mostly looks good to me 👍 
Going to let @ogaca-dd have a look before we merge.

ogaca-dd reviewed (changes requested):
@AndrewMD5 thanks a lot for all the improvements 👍 

There are a lot of changes in this PR and I have just started the review (more to come).

Can you explain why you added `CancellationToken`?
Can you also rebase your PR from `origin/master` to hide change related to https://github.com/DataDog/serilog-sinks-datadog-logs/pull/68?

AndrewMD5 commented on src/Serilog.Sinks.Datadog.Logs/Sinks/Datadog/DatadogSink.cs:
```diff
@@ -93,42 +126,57 @@ protected override async Task EmitBatchAsync(IEnumerable<LogEvent> events)
+                try
+                {
+                    // delay the dispose by one batch period so lingering events get logged. 
+                    Task.Delay(DefaultBatchPeriod, _cancellationToken).Wait(_cancellationToken);
```
The reason for this delay is that if you use the sink as shown in README.md:
```
using (var log = new LoggerConfiguration()
    .WriteTo.DatadogLogs("<API_KEY>")
    .CreateLogger())
{
    log.Information("Processed {@Position} in {Elapsed:000} ms.", position, elapsedMs);
}
```

The call to ` log.Information` will complete immediately and the using block will exit which raises `Dispose()` - because we now clean up the sink resources and flow the `CancellationToken`, the event passed to `log.Information` will never reach the wire. You're basically writing to `/dev/null` at that point.

So, to avoid breaking existing implementations _and_ to allow the sink a chance to write pending batches this block makes sense. Users also shouldn't be constantly disposing the sink over-and-over so this block is really only ever going to be called at the end of the application. ` CloseAndFlush` also does not pump any pending events to the `WriteAsync` method, so really this block is purely about giving things in-flight a chance to finish.

AndrewMD5 commented on src/Serilog.Sinks.Datadog.Logs/Sinks/Datadog/DatadogHttpClient.cs:
```diff
@@ -36,123 +42,141 @@ public class DatadogHttpClient : IDatadogClient
+                        break;
+                    }
+                    var formattedLog = _formatter.FormatMessage(events[i]);
+                    var logSize = Encoding.UTF8.GetMaxByteCount(formattedLog.Length);
```
I'll switch this back. Because we work primarily with ASCII data `GetMaxByteCount` is usually safe, but I realize that isn't going to be the case for everyone.

AndrewMD5 commented:
> Can you explain why you added `CancellationToken`?

Cancellation is cooperative in .NET. Everything in the call-chain has to be explicitly passed the CancellationToken in order for it to work well. This means you need to explicitly pass the token into other APIs that take a token if you want cancellation to be most effective. So to dispose the sink, we have to dispose all the resources the sink created, and cancel any task those resources might be currently executing. This also helps avoid runtime exceptions.

AndrewMD5 commented:
> Can you also rebase your PR from `origin/master` to hide change related to #68?

IIRC it already is based on `origin/master`; do you mean since #68 is merged in already that you'd prefer if I rebase off 4c0dab81dd2b6e1a7f6ddbfc385f86ee90501644 to hide the changes #68 made?

ogaca-dd commented:
@AndrewMD5: Thank you for your quick feedback. 
As the PR has a lot of changes and as I have a busy week, I cannot review the PR this week, sorry for that. I will review this PR next week.

AndrewMD5 commented:
> @AndrewMD5: Thank you for your quick feedback. As the PR has a lot of changes and as I have a busy week, I cannot review the PR this week, sorry for that. I will review this PR next week.

Do you think you'll be able to get this reviewed this week?

ogaca-dd reviewed (changes requested):
Thanks again for the contribution.
The code looks great.
Let me know if you have any questions about my comments.

@gh123man : I did not review src/Serilog.Sinks.Datadog.Logs/Sinks/Datadog/DatadogFormatter.cs but I think this code should be reviewed.

AndrewMD5 commented:
Happy new year folks! I’ll address the comments this weekend and get any necessary changes in.

jszwedko commented:
Closing due to lack of response, but feel free to reopen if you want to pick this back up! Thanks for the initial efforts here!
